book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,335 @@
+
+Keep the data dependencies and checks as simple as possible.
+
+For example, Zebra could defer checking both the BlockHeight and Network Upgrade.
+
+But since the Network Upgrade depends on the BlockHeight, we only need to defer
+the BlockHeight check. Then we can use all the fields that depend on the
+BlockHeight, as if it is correct. If the final BlockHeight check fails, we will
+reject the entire block, including all the verification we perfomed using the
+assumed Network Upgrade.
+
+### Implementation Strategy
+[implementation-strategy]: #implementation-strategy
+
+When implementing these designs, perform as much verification as possible, await
+any dependencies, then perform the necessary checks.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Verification Stages
+[verification-stages]: #verification-stages
+
+In Zebra, verification occurs in the following stages:
+* **Structural Verification:** Raw block data is parsed into a block header and
+  transactions. Invalid data is not representable in these structures:
+  deserialization (parsing) can fail, but serialization always succeeds.
+* **Semantic Verification:** Parsed block fields are verified, based on their
+  data dependencies:
+  * Context-free fields have no data dependencies, so they can be verified as
+    needed.
+  * Fields with simple data dependencies defer that dependency as long as
+    possible, so they can perform more verification in parallel. Then they await
+    the required data, which is typically the previous block. (And potentially
+    older blocks in its chain fork.)
+  * Fields with complex data dependencies require their own parallel verification
+    designs. These designs are out of scope for this RFC.
+* **State Updates:** After a block is verified, it is added to the state. The
+  details of state updates, and their interaction with semantic verification,
+  are out of scope for this RFC.
+
+This RFC focuses on Semantic Verification, and the design patterns that enable
+blocks to be verified in parallel.
+
+## Verification Interfaces
+[verification-interfaces]: #verification-interfaces
+
+Verification is implemented by the following traits and services:
+* **Structural Verification:**
+  * `zebra_network::init`: Provides a downloader service that accepts a
+    `BlockHeaderHash` request, and parses the peer response into a `Block`
+    struct.
+  * `zebra_chain::ZcashDeserialize`: A trait for parsing consensus-critical
+    data structures from a byte buffer.
+* **Semantic Verification:**
+  * `ChainVerifier`: Provides a verifier service that accepts a `Block` request,
+    performs verification on the block, and responds with a `BlockHeaderHash` on
+    success.
+  * Internally, the `ChainVerifier` selects between a `CheckpointVerifier` for
+    blocks that are within the checkpoint range, and a `BlockVerifier` for
+    recent blocks.
+* **State Updates:**
```

Comment:
same as https://github.com/ZcashFoundation/zebra/pull/763/files#r470200712

Reply:
Fixed in fb53bb8b7f45e8659b0e1f2d35ad8e01f1577b5c.